x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PruneColumns.java

```diff
@@ -131,14 +139,15 @@ public LogicalPlan apply(LogicalPlan plan) {
      * Prunes attributes from the list not found in the given set.
      * Returns null if no changed occurred.
      */
-    private static <N extends NamedExpression> List<N> removeUnused(List<N> named, AttributeSet.Builder used) {
+    private static <N extends NamedExpression> List<N> removeUnused(List<N> named, AttributeSet.Builder used, List<Attribute> exceptions) {
         var clone = new ArrayList<>(named);
         var it = clone.listIterator(clone.size());
 
         // due to Eval, go in reverse
         while (it.hasPrevious()) {
             N prev = it.previous();
-            if (used.contains(prev.toAttribute()) == false) {
+            var attr = prev.toAttribute();
+            if (used.contains(attr) == false && exceptions.contains(attr) == false) {
```

Comment:
This was preventing queries like https://github.com/elastic/elasticsearch/pull/128917/files#diff-a5ed793b796c14ec1f92807242ff2376e16888c8cf7627a4f6a8f57f23ed843aR1061-R1115.